pkg/jobworker/tail.go

```diff
@@ -0,0 +1,40 @@
+package jobworker
+
+import (
+	"io"
+	"os"
+	"time"
+)
+
+// tailReader wraps the normal io.ReadCloser with an implementation that sleeps for a specified pollInterval
+// before retrying to Read
+type tailReader struct {
+	io.ReadCloser
+	pollInterval time.Duration
+}
+
+// Read calls the normal io.ReadCloser and checks for an io.EOF error and skips returning, to sleep instead
+func (t tailReader) Read(b []byte) (int, error) {
+	for {
+		n, err := t.ReadCloser.Read(b)
+		if n > 0 {
+			return n, nil
+		} else if err != io.EOF {
+			return n, err
+		}
+		time.Sleep(t.pollInterval)
+	}
+}
+
+// newTailReader opens a file by path, sets the read offset to the start and returns a wrapped tailReader to the file
+func newTailReader(fileName string, pollInterval time.Duration) (tailReader, error) {
+	f, err := os.Open(fileName)
```

Comment:
That's correct, I went with this approach for it's simplicity and convenience to the caller. Since I am returning an io.ReadCloser the caller can read the log output as they like. My alternative approaches always involved having to keep the entire log output in memory so that I could return a new buffer to read from each time, for multiple concurrent readers, or using channels to "fan out" each log line, where concurrent readers could then read the lines from the channel. With the os.File approach each reader only puts what it is currently reading into memory.

In the end though I chose this approach because the number of open file descriptors, being the biggest drawback, can be configured on most systems to a large number. For example setting up something like nginx normally involves increasing this limit, which can be set to 100,000 on most systems, example https://www.cyberciti.biz/faq/linux-increase-the-maximum-number-of-open-files/.

I also mentioned this limitation in the design doc.

> No set up has occurred on the host running the server so performance issues would be expected if run at scale. Often linux will have restrictions on the number of processes or open file descriptors a user can have. Those two come to mind immediately but load testing would need to be performed to likely tune other parameters on the system.

https://github.com/bkneis/teleport-jobworker/blob/main/0001-job-worker.md#trade-offs--limitations